langstream-api-gateway/src/main/java/ai/langstream/apigateway/gateways/ConsumeGateway.java

```diff
@@ -0,0 +1,224 @@
+                    } catch (InterruptedException | CancellationException ex) {
+                        // ignore
+                    } catch (Throwable ex) {
+                        log.error(ex.getMessage(), ex);
+                    } finally {
+                        closeReader();
+                    }
+                },
+                executor);
+    }
+
+    protected void readMessages(Supplier<Boolean> stop, Consumer<String> onMessage)
+            throws Exception {
+        while (true) {
+            if (Thread.currentThread().isInterrupted()) {
+                return;
+            }
+            if (stop.get()) {
+                return;
+            }
+            final TopicReadResult readResult = reader.read();
+            final List<Record> records = readResult.records();
+            for (Record record : records) {
+                log.debug("Received record {}", record);
+                boolean skip = false;
+                if (filters != null) {
+                    for (Function<Record, Boolean> filter : filters) {
+                        if (!filter.apply(record)) {
+                            skip = true;
+                            log.debug("Skipping record {}", requestContext, record);
+                            break;
+                        }
+                    }
+                }
+                if (!skip) {
+                    final Map<String, String> messageHeaders = computeMessageHeaders(record);
+                    final String offset = computeOffset(readResult);
+
+                    final ConsumePushMessage message =
+                            new ConsumePushMessage(
+                                    new ConsumePushMessage.Record(
+                                            record.key(), record.value(), messageHeaders),
+                                    offset);
+                    final String jsonMessage = mapper.writeValueAsString(message);
+                    onMessage.accept(jsonMessage);
+                }
+            }
+        }
+    }
+
+    private static Map<String, String> computeMessageHeaders(Record record) {
+        final Collection<Header> headers = record.headers();
+        final Map<String, String> messageHeaders;
+        if (headers == null) {
+            messageHeaders = Map.of();
+        } else {
+            messageHeaders = new HashMap<>();
+            headers.forEach(h -> messageHeaders.put(h.key(), h.valueAsString()));
+        }
+        return messageHeaders;
+    }
+
+    private static String computeOffset(TopicReadResult readResult) {
+        final byte[] offset = readResult.offset();
+        if (offset == null) {
+            return null;
+        }
+        return Base64.getEncoder().encodeToString(offset);
+    }
+
+    private void closeReader() {
+        if (reader != null) {
+            try {
+                reader.close();
+            } catch (Exception e) {
+                log.error("error closing reader", e);
```

Comment:
it is better to report the topic name, or anything that connect this error with an application/tenant.... otherwise it is almost useless

Reply:
usually there's the kafka/pulsar client errors logged with the name of the topic before this error. anyway, this error is not that important, I'll move it to debug